natlib/svcd.c

```diff
@@ -185,3 +187,84 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+// Lua: storm.n.svcd_add_service ( svcd_id )
+// Add a new service to the service daemon
+// this must be called before SVCD.advertise()
+static int svcd_add_service( lua_State *L ) {
+
+    //TODO: Garbage collection?
+
+    lua_getglobal(L, "SVCD");
+
+    lua_pushstring(L, "blsmap");
+    lua_gettable(L, 2);
+    lua_pushvalue(L, 1); //svc_id @ index 4
+    lua_pushlightfunction(L, libstorm_bl_addservice);
+    lua_pushvalue(L, 1); //svc_id @ index 6
+    lua_call(L, 1, 1);
+    lua_settable(L, 3);
+
+    lua_pushstring(L, "blamap");
+    lua_gettable(L, 2);
+    lua_pushvalue(L, 1); //svc_id @ index 5
+    lua_newtable(L);
+    lua_settable(L, 4);
+
+    lua_pushstring(L, "manifest");
+    lua_gettable(L, 2);
+    lua_pushvalue(L, 1); //svc_id @ index 6
+    lua_newtable(L);
+    lua_settable(L, 5);
+
+    lua_pushstring(L, "manifest_map");
+    lua_gettable(L, 2);
+    lua_pushvalue(L, 1); //svc_id @ index 7
+    lua_newtable(L);
+    lua_settable(L, 6);
+
+    return 0;
+}
+
+// Lua: storm.n.svcd_add_attribute ( svc_id, attr_id, write_fn )
+// Add a new attribute to a service in the service daemon
+static int svcd_add_attribute( lua_State *L ) {
+
+    lua_getglobal(L, "SVCD"); //Index 4
+
+    lua_pushstring(L, "blsmap"); //Index 5
+    lua_gettable(L, 4);
+
```

Comment:
My bad. I didn't see that the function took in three attributes and that they'd be on the stack. I yet feel that the stack is addressed opposite to what you have done though.